### PR TITLE
Removes References To Built-in Custom Data Implementations

### PR DESCRIPTION
**Pull request by gsalaz98**
#### Description

Continuation of: #5785 

This PR removes references to custom data types that have been added directly into the Lean repo over the course of the last two years.

Some new "Iconic" custom data types were added to keep some of the important Lean tests related to custom data. These data types are mocked versions of actual data types, containing functionality that was previously present in certain data types.

The new recommended pattern of adding data types to be supported in the cloud is via the instructions contained within the https://github.com/QuantConnect/Lean.DataSource.SDK repo

Data types removed & their new repos:
  * https://github.com/QuantConnect/Lean.DataSource.SEC (`SECReport8K, SECReport10K, SECReport10Q`)
  * https://github.com/QuantConnect/Lean.DataSource.TradingEconomics (`TradingEconomicsCalendar, TradingEconomicsIndicator`)
  * https://github.com/QuantConnect/Lean.DataSource.Estimize (`EstimizeConsensus, EstimizeRelease, EstimizeEstimate`)
  * https://github.com/QuantConnect/Lean.DataSource.SmartInsiderIntentionsTransactions (`SmartInsiderIntention, SmartInsiderTransaction`)
  * https://github.com/QuantConnect/Lean.DataSource.BenzingaNews (`BenzingaNews`)
  * https://github.com/QuantConnect/Lean.DataSource.USTreasury (`USTreasuryYieldCurveRate`)
  * https://github.com/QuantConnect/Lean.DataSource.USEnergy (`USEnergy`)
  * https://github.com/QuantConnect/Lean.DataSource.FRED (`Fred`)
  * https://github.com/QuantConnect/Lean.DataSource.TiingoNews (`TiingoNews`)
  * https://github.com/QuantConnect/Lean.DataSource.CBOE (`CBOE`)
  * https://github.com/QuantConnect/Lean.DataSource.QuiverQuantWikipedia (`QuiverWikipedia`)
  * https://github.com/QuantConnect/Lean.DataSource.QuiverQuantWallStreetBets (`QuiverWallStreetBets`)
  * https://github.com/QuantConnect/Lean.DataSource.QuiverQuantEventsBeta (`QuiverEventsBeta`)
  * https://github.com/QuantConnect/Lean.DataSource.QuiverQuantCongressTrading (`QuiverCongress`)
  * https://github.com/QuantConnect/Lean.DataSource.VIXCentral (`VIXCentral`)

:warning: This is a work in progress, since there's some more references we have to remove in other repos, so please don't merge yet.
#### Related Issue

N/A
#### Motivation and Context

Removes class name clashes and unifies implementations of the same data types going forward, without the need to update this repo; Allows data vendors to implement their own data types without having to merge directly into this repo.
#### Requires Documentation Change

Yes.

Instead of importing from the `QuantConnect.Data.Custom.*` namespaces, we should direct users to import from `QuantConnect.DataSource.*`
#### How Has This Been Tested?

All unit tests passing, cloud has been working for a few weeks with new data types when not importing the data from the Custom.* namespace
#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [ ] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Algorithm.Python/SmartInsiderDataAlgorithm.py [hunk 1]
```diff
@@ -1,64 +0,0 @@
...
-    def Initialize(self):
-        '''Initialise the data and resolution required, as well as the cash and start-end dates for your algorithm. All algorithms must initialized.'''
-
-        self.SetStartDate(2019, 7, 25)
-        self.SetEndDate(2019, 8, 2)
-        self.SetCash(100000)
-
-        self.symbol = self.AddEquity("KO", Resolution.Daily).Symbol
-        self.AddData(SmartInsiderTransaction, "KO")
-        self.AddData(SmartInsiderIntention, "KO")
```
Comment (Martin-Molinero) [inline Algorithm.Python/SmartInsiderDataAlgorithm.py:34]: Repo at https://github.com/QuantConnect/Lean.DataSource.SmartInsiderIntentionsTransactions only has `Demonstration.cs` think we should keep both `.py` and `.cs` example algorithms

Comment (Martin-Molinero) [reply Algorithm.Python/SmartInsiderDataAlgorithm.py:34]: Same for all repos 💯

File: Engine/HistoricalData/SubscriptionDataReaderHistoryProvider.cs [hunk 2]
```diff
@@ -42,7 +42,7 @@ public class SubscriptionDataReaderHistoryProvider : SynchronizingHistoryProvide
         private IMapFileProvider _mapFileProvider;
         private IFactorFileProvider _factorFileProvider;
         private IDataCacheProvider _dataCacheProvider;
-        private IDataPermissionManager _dataPermissionManager;
+        protected IDataPermissionManager _dataPermissionManager;
```
Comment (Martin-Molinero) [inline Engine/HistoricalData/SubscriptionDataReaderHistoryProvider.cs:45]: nit - `_dataPermissionManager` -> `DataPermissionManager` + xml docs

File: Common/Data/Custom/IconicTypes/LinkedData.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+using QuantConnect.Data;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using ProtoBuf;
+
+namespace QuantConnect.Data.Custom.IconicTypes
+{
+    /// <summary>
+    /// Data source that is unlinked (no mapping) and takes any ticker when calling AddData
```
Comment (Martin-Molinero) [inline Common/Data/Custom/IconicTypes/LinkedData.cs:26]: nit - docs are of Unlinked version

File: Common/Data/Custom/IconicTypes/IndexedLinkedData2.cs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+using ProtoBuf;
+
+namespace QuantConnect.Data.Custom.IconicTypes 
+{
+    /// <summary>
+    /// Data type that is indexed, i.e. a file that points to another file containing the contents
+    /// we're looking for in a Symbol.
+    /// </summary>
+    [ProtoContract(SkipConstructor = true)]
+    public class IndexedLinkedData2 : IndexedBaseData
```
Comment (Martin-Molinero) [inline Common/Data/Custom/IconicTypes/IndexedLinkedData2.cs:30]: What's the difference with `IndexedLinkedData`?

File: Algorithm.CSharp/DynamicSecurityDataAlgorithm.cs [hunk 5]
```diff
@@ -1,71 +0,0 @@
...
-            SECReport10K googlSec10kReport = GOOGL.Data.Get<SECReport10K>();
-            Log($"{Time:o}:  8K: {googlSec8kReport}");
-            Log($"{Time:o}: 10K: {googlSec10kReport}");
-
-            // 2. Get the list of data points of a particular type for the most recent time step:
-            // 2.a Using the C# generic method, GetAll<T>:
-            List<SECReport8K> googlSec8kReports = GOOGL.Data.GetAll<SECReport8K>();
-            List<SECReport10K> googlSec10kReports = GOOGL.Data.GetAll<SECReport10K>();
-            Log($"{Time:o}: List:  8K: {googlSec8kReports.Count}");
-            Log($"{Time:o}: List: 10K: {googlSec10kReports.Count}");
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/DynamicSecurityDataAlgorithm.cs:63]: Think we need to re add this algorithm and make him part of the regression test 👍, same for the Py version, can just use some of the Iconic data types

**Review by Martin-Molinero — APPROVED** [review]
Looks good, leaving a few requests 👍

File: Common/Data/Custom/IconicTypes/IndexedLinkedData2.cs (hunk 4 above)
Comment (gsalaz98) [reply Common/Data/Custom/IconicTypes/IndexedLinkedData2.cs:30]: `IndexedLinkedData2` throws whenever `GetSource` is called with `isLiveMode = true`,  whereas the other doesn't. This behavior is required to keep parity with some tests that were previously Benzinga/Tiingo news data sources.

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**